### PR TITLE
Emit export declarations in place

### LINKED ISSUE #4101: Replacing object reference before exporting causing problems
Hi,

If I have the following typescript code:
```javascript
var obj = {
    someKey: 6
};

// For some reason I want to change the object
obj = {
    someKey: 7
};

export {obj};
```

it compiles to:
```javascript
var obj = {
    someKey: 6
};
exports.obj = obj;
obj = {
    someKey: 7
};
```

So if I importing the object I get:
```javascript
import {obj} from './some-file';
console.log(obj.someKey === 6); // true
```

But I was expecting `7`. I.e. I was expecting the following JS code:
```javascript
var obj = {
    someKey: 6
};
obj = {
    someKey: 7
};
exports.obj = obj;
```

Is this a bug or expected behavior?

I am using version `1.6.0-dev.20150731` but it has the same behavior in `1.5.3`.

### PR DESCRIPTION
**Pull request by sheetalkamat**
Fixes #4101 and #4004

### CODE REVIEW

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -6145,42 +6112,42 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
...
-                                    // emit re-exports for named imports
-                                    // import {a, b} from 'foo'
-                                    // export {a, b as c}
-                                    for (let element of (<NamedImports>namedBindings).elements) {
-                                        emitExportMemberAssignments(element.name || element.propertyName);
-                                        writeLine();
-                                    }
-                                }
-                            }
+                            //let defaultName =
```
Comment (mhegazy) [inline src/compiler/emitter.ts]: should #4179 go in first?

File: tests/baselines/reference/exportSpecifierForAGlobal.js [hunk 2]
```diff
@@ -13,6 +13,7 @@ export function f() {
 
 
 //// [b.js]
+exports.X = X;
 function f() {
```
Comment (DickvdBrink) [inline tests/baselines/reference/exportSpecifierForAGlobal.js]: This export wasn't there previously; correct?

Comment (mhegazy) [inline tests/baselines/reference/exportSpecifierForAGlobal.js]: the issue is reported in #4004

Comment (DickvdBrink) [inline tests/baselines/reference/exportSpecifierForAGlobal.js]: Ah, sorry about that :)

**Comment by ahejlsberg** [comment]
Isn't the following now broken?

```typescript
export { foo };
var foo = 1;
```

The reason we emit the export assignment right after the variable declaration is that the export declaration might occur anywhere in the file. Also, there might be more than one export declaration for a particular variable (e.g. to export the same variable by different names), but ideally we only want to generate one export assignment.

**Comment by sheetalkamat** [comment]
@ahejlsberg : as @mhegazy pointed out wouldn't it be easier to fix this since user has workaround for it than the bug its fixing right now. There is no work around for the current scenario.

**Comment by mhegazy** [comment]
This should not be needed any longer.

**Pull request closed without merging**